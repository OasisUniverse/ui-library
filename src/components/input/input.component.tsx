import React, { ChangeEvent, FC, ForwardedRef, forwardRef, useCallback, useMemo, useState } from 'react';
import { DebounceInput } from 'react-debounce-input';
import InputCrossIcon from 'src/shared/svg/input-cross.icon';
import styles from './input.module.scss';
import ClosedEyeIcon from '../../shared/svg/closed-eye.icon';
import OpenedEyeIcon from '../../shared/svg/opened-eye.icon';
import LockIcon from '../../shared/svg/lock.icon';

const INPUT_TYPE = 'text';
const INPUT_PASSWORD_TYPE = 'password';

export enum InputType {
    DefaultInput,
    PasswordInput,
    SearchInput,
    TextArea,
}

export enum LabelPosition {
    TopLeft = 'topLeft',
    TopRight = 'topRight',
    BottomLeft = 'bottomLeft',
    BottomRight = 'bottomRight',
}

export interface InputProps {
    value?: string;
    onChange: (value: string, event?: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
    type?: InputType;
    placeholder?: string;
    autocomplete?: boolean;
    className?: string;
    debounceTimeoutMs?: number;
    isError?: boolean;
    label?: string;
    labelPosition?: LabelPosition;
    isResizable?: boolean;
    autofocus?: boolean;
    disabled?: boolean;
}

export const Input: FC<InputProps> = forwardRef(
    (
        {
            type = InputType.DefaultInput,
            onChange,
            label,
            labelPosition = LabelPosition.TopRight,
            debounceTimeoutMs,
            className = '',
            isResizable = false,
            autofocus = false,
            value = '',
            placeholder = '',
            autocomplete = false,
            isError = false,
            disabled = false,
        },
        forwardRef,
    ) => {
        const [isDisplayPassword, setDisplayPassword] = useState<boolean>(false);
        const isRenderPasswordIcon = useMemo(
            () => type === InputType.PasswordInput && !disabled && value,
            [type, disabled, value],
        );
        const displayPasswordHandler = () => setDisplayPassword(true);
        const hidingPasswordHandler = () => setDisplayPassword(false);

        const labelElement = label && (
            <span className={`${styles.label} ${labelPosition && styles[labelPosition]}`}>{label}</span>
        );

        const clearInput = () => onChange('');

        const onChangeCallback = useCallback(
            (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => onChange(event.target.value, event),
            [onChange],
        );

        const isRenderCrossIcon = useMemo(
            () => value && !disabled && type !== InputType.TextArea && type !== InputType.PasswordInput,
            [type, value, disabled],
        );
        const crossIcon = isRenderCrossIcon && (
            <span className={styles.cross} onClick={clearInput}>
                <InputCrossIcon />
            </span>
        );

        const isActiveStyle = useMemo(() => (value && !disabled ? styles.active : ''), [value, disabled]);
        const isErrorStyle = useMemo(() => (isError && !disabled ? styles.error : ''), [isError, disabled]);
        const isDisabled = useMemo(() => (disabled ? styles.disabled : ''), [disabled]);
        const isResizableStyle = useMemo(() => (isResizable ? styles.resize : ''), [isResizable]);

        const inputWrapperClassName = useMemo(
            () =>
                `${type !== InputType.TextArea ? styles.inputWrapper : ''} ${
                    type === InputType.SearchInput ? styles.searchWrapper : ''
                } ${value ? isActiveStyle : ''} ${isError ? isErrorStyle : ''} ${isDisabled}`,
            [type, value, isError],
        );
        return (
            <div className={`${styles.wrapper} ${className}`}>
                {labelElement}
                {type === InputType.TextArea ? (
                    <textarea
                        className={`${isActiveStyle} ${isDisabled} ${isErrorStyle} ${isResizableStyle}`}
                        autoComplete={autocomplete ? 'on' : 'off'}
                        placeholder={placeholder}
                        ref={forwardRef as ForwardedRef<HTMLTextAreaElement>}
                        autoFocus={autofocus}
                        value={value}
                        onChange={onChangeCallback}
                    />
                ) : (
                    <div className={inputWrapperClassName}>
                        <DebounceInput
                            type={
                                type === InputType.PasswordInput && !isDisplayPassword
                                    ? INPUT_PASSWORD_TYPE
                                    : INPUT_TYPE
                            }
                            ref={forwardRef as ForwardedRef<DebounceInput<HTMLInputElement>>}
                            disabled={disabled}
                            autoComplete={autocomplete ? 'on' : 'off'}
                            placeholder={placeholder}
                            autoFocus={autofocus}
                            debounceTimeout={debounceTimeoutMs}
                            value={value}
                            onChange={onChangeCallback}
                        />
                        {disabled && <LockIcon />}
                        {crossIcon}
                        {isRenderPasswordIcon && (
                            <span
                                className={styles.passwordDisplayButton}
                                onPointerDown={displayPasswordHandler}
                                onPointerUp={hidingPasswordHandler}
                            >
                                {isDisplayPassword ? <OpenedEyeIcon /> : <ClosedEyeIcon />}
                            </span>
                        )}
                    </div>
                )}
            </div>
        );
    },
);

Input.displayName = 'Input';
