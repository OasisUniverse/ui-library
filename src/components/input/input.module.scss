@import '../../styles/colors';
@import '../../styles/fonts';

.wrapper {
	position: relative;
	display: flex;
	font-family: Inter, sans-serif;
	
	.label {
		font-family: inherit;
		position: absolute;
		font-size: 12px;
		color: white;
		
		&.topLeft {
			left: 0;
			top: -18px;
		}
		
		&.topRight {
			right: 0;
			top: -18px;
		}
		
		&.bottomLeft {
			left: 0;
			bottom: -18px;
		}
		
		&.bottomRight {
			right: 0;
			bottom: -18px;
		}
	}
	
	.cross {
		display: flex;
		align-items: center;
		margin-left: auto;
		cursor: pointer;
	}
	
	.inputWrapper,
	textarea {
		width: 100%;
		border-radius: 16px;
		border: 2px solid color(primary, dark);
		box-sizing: border-box;
		background-color: color(background, content);
		transition: 0.6s ease;
		transition-delay: 0.3s;
		transition-property: color, background, border, box-shadow;
		
		&.active {
			background: color(background, main);
			border: 2px solid color(accent, dark);
			box-shadow: 0 0 15px color(accent, dark), inset 0 0 10px color(accent, dark);
			color: white;
			
			&::placeholder {
				color: white;
			}
			
			&.error {
				border: 2px solid color(statuses, error);
				box-shadow: 0 0 15px color(statuses, error), inset 0 0 10px color(statuses, error);
			}
		}
		
		&.error {
			background: color(background, main);
			border: 2px solid color(statuses, error);
			box-shadow: 0 0 15px color(statuses, error), inset 0 0 10px color(statuses, error);
			color: white;
			
			&::placeholder {
				color: white;
			}
		}
		
		&.resize {
			resize: vertical;
		}
	}
	
	textarea {
		resize: none;
		padding: 10px;
		outline: none;
		overflow-y: auto;
		height: 100%;
		font-size: 16px;
		font-family: inherit;
		scroll-snap-type: y proximity;
		
		&::-webkit-scrollbar-track {
			width: 0;
		}
		
		&::-webkit-scrollbar {
			width: 0;
		}
		
		&::-webkit-scrollbar-thumb {
			width: 0;
		}
		
		&::placeholder {
			transition: all 0.3s ease;
			color: color(inneres, placeholder);
		}
		
		&:hover {
			border: 2px solid color(accent, dark);
			background-color: color(background, content);
			box-shadow: inset 0 0 10px color(accent, dark);
			
			&::placeholder {
				color: color(accent, dark);
			}
			
			&.active {
				background: color(background, main);
				color: white;
			}
		}
		
		&:disabled {
			background: color(background, main);
			border-color: color(statuses, muted);
		}
	}
	
	.inputWrapper {
		position: relative;
		display: flex;
		align-items: center;
		height: 64px;
		color: white;
		padding: 0 24px 0 24px;
		
		&.passwordInput {
			padding: 0 12px 0 24px;
		}
		
		&.active,
		&.error,
		&.disabled {
			background: color(background, main);
		}
		
		.passwordDisplayButton {
			display: flex;
			align-items: center;
			height: 22px;
			width: 22px;
			cursor: pointer;
		}
		
		input {
			width: 100%;
			display: flex;
			align-items: center;
			height: 100%;
			box-sizing: border-box;
			border: none;
			outline: none;
			border-radius: inherit;
			text-overflow: ellipsis;
			transition: 0.3s ease-in color;
			font-size: 16px;
			font-family: inherit;
			background: transparent;
			
			&::placeholder {
				transition: all 0.3s ease;
				color: color(inneres, placeholder);
			}
		}
		
		&.disabled {
			border-color: color(statuses, muted);
			
			input {
				font-weight: 500;
				color: color(inneres, placeholder);
			}
			
			input::placeholder {
				font-weight: 400;
				color: color(statuses, muted);
			}
		}
		
		&.active input,
		&.error input {
			color: white;
			background-color: transparent;
		}
		
		&.active {
			border: 2px solid color(accent, dark);
			box-shadow: 0 0 15px color(accent, dark), inset 0 0 10px color(accent, dark);
			
			.elementWrapper {
				border-left: 1px solid color(accent, dark);
				
				&:nth-child(2) {
					border: none;
				}
			}
			
			&:hover {
				box-shadow: 0 0 15px color(accent, dark), inset 0 0 10px color(accent, dark);
			}
		}
		
		&.error {
			border: 2px solid color(statuses, error);
			box-shadow: 0 0 15px color(statuses, error), inset 0 0 10px color(statuses, error);
		}
		
		&.searchWrapper {
			border-radius: 25px;
			
			&.active {
				border: 2px solid color(accent, dark);
				box-shadow: 0 0 15px color(accent, dark), inset 0 0 10px color(accent, dark);
			}
		}
		
		&:hover {
			border: 2px solid color(accent, dark);
			box-shadow: inset 0 0 10px color(accent, dark);
			color: color(accent, dark);
			
			input::placeholder {
				color: color(accent, dark);
			}
			
			&.error {
				border: 2px solid color(statuses, error);
				box-shadow: 0 0 15px color(statuses, error), inset 0 0 10px color(statuses, error);
				
				input::placeholder {
					color: white;
				}
			}
			
			&.disabled {
				box-shadow: none;
				color: color(statuses, muted);
				border-color: color(statuses, muted);
				cursor: not-allowed;
				
				input {
					cursor: not-allowed;
					
					&::placeholder {
						color: color(statuses, muted);
					}
				}
			}
		}
	}
}
