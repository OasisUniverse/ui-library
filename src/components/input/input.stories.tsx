import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { Input, InputProps } from './input.component';
import { defaultInputProps, searchInputProps, textInputProps } from '../../__mocks__/input.mock';

export default {
    title: 'Input',
    component: Input,
} as Meta;

const Template: Story<InputProps> = (args: InputProps) => <Input {...args} />;

export const EventType = Template.bind({});

export const Presentation = (): JSX.Element => (
    <div style={{ background: '#1a1a22', padding: '50px' }}>
        <span style={{ fontSize: '30px', marginBottom: '15px', color: 'white' }}>Input</span>
        <div
            style={{
                boxSizing: 'border-box',
                display: 'grid',
                gridTemplateRows: 'auto',
                gridTemplateColumns: '1fr 1fr',
                gridGap: '20px',
            }}
        >
            {defaultInputProps.map((args, index) => (
                <div style={{ marginBottom: '25px', boxSizing: 'border-box' }} key={index}>
                    <Input {...args} />
                </div>
            ))}
        </div>
        <span style={{ fontSize: '30px', marginBottom: '15px' }}>SearchInput</span>
        <div
            style={{
                boxSizing: 'border-box',
                display: 'grid',
                gridTemplateRows: 'auto',
                gridTemplateColumns: '1fr 1fr',
                gridGap: '20px',
            }}
        >
            {searchInputProps.map((args, index) => (
                <div style={{ marginTop: '12.5px', marginBottom: '12.5px' }} key={index}>
                    <Input {...args} />
                </div>
            ))}
        </div>
        <span style={{ fontSize: '30px' }}>TextArea</span>
        <div
            style={{
                width: '100%',
                boxSizing: 'border-box',
                marginLeft: '-20px',
                display: 'grid',
                gridTemplateRows: 'auto',
                gridTemplateColumns: '1fr 1fr',
                gridGap: '20px',
            }}
        >
            {textInputProps.map((args, index) => (
                <div
                    style={{
                        marginTop: '12.5px',
                        marginBottom: '12.5px',
                        marginRight: '15px',
                        width: '100%',
                        padding: '5px 25px',
                    }}
                    key={index}
                >
                    <Input {...args} />
                </div>
            ))}
        </div>
    </div>
);

EventType.args = {
    value: '',
    onChange: (): void => {
        return;
    },

    placeholder: 'Введите текст...',
    autocomplete: true,
};
