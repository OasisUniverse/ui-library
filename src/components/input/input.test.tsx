import React from 'react';
import { Input, InputProps } from './input.component';
import Enzyme, { shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { inputTestMockMap } from '../../__mocks__/input.test.mock';

Enzyme.configure({ adapter: new Adapter() });

const defaultInputProps: InputProps = {
    onChange: jest.fn(),
};

let defaultInput: ShallowWrapper;

describe('<Input /> test list', () => {
    beforeAll(() => {
        defaultInput = shallow(<Input {...defaultInputProps} />);
        jest.clearAllMocks();
    });
    it('render all list', () => {
        Array.from(inputTestMockMap, ([key, props]) => {
            const component = shallow(<Input key={key} {...props} />);
            expect(component.exists()).toBeTruthy();
            expect(component).toMatchSnapshot();
        });
    });
    it('should return value via onChange', () => {
        const input = defaultInput.find('.inputWrapper').childAt(0);
        input.simulate('change', { target: { value: 'text' } });
        expect(defaultInputProps.onChange).toHaveBeenCalledWith('text', { target: { value: 'text' } });
    });
    it('Has a label and label has position="bottomLeft"', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('label-position')} />);
        expect(defaultInput.find('.label').exists()).toBeTruthy();
        expect(defaultInput.find('.label').hasClass('bottomLeft')).toBeTruthy();
    });
    it('has a placeholder', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-has-placeholder')} />);
        const inputWrapper = defaultInput.find('.inputWrapper');
        const input = inputWrapper.find('DebounceInput');
        expect(input.props().placeholder).toEqual('test');
    });
    it('has autofocus', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-has-autofocus')} />);
        const inputWrapper = defaultInput.find('.inputWrapper');
        const input = inputWrapper.find('DebounceInput');
        expect(input.props().autoFocus).toEqual(true);
    });
    it('has catches an error', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-catches-error')} />);
        const inputWrapper = defaultInput.find('.inputWrapper');
        expect(inputWrapper.hasClass('error')).toBeTruthy();
    });
    it('disabled', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-disabled')} />);
        const inputWrapper = defaultInput.find('.inputWrapper');
        const input = inputWrapper.find('DebounceInput');
        expect(input.props().disabled).toEqual(true);
        expect(inputWrapper.hasClass('disabled')).toBeTruthy();
    });
    it('has a cross icon', () => {
        const defaultInput = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-has-cross-icon')} />);
        const inputWrapper = defaultInput.find('.inputWrapper');
        const input = inputWrapper.find('DebounceInput');
        expect(input.props().value).toEqual('test');
        expect(inputWrapper.find('.cross').exists()).toBeTruthy();
    });
    it('resizable', () => {
        const textAreaComponent = shallow(<Input {...defaultInputProps} {...inputTestMockMap.get('it-resizable')} />);
        const textarea = textAreaComponent.find('textarea');
        expect(textarea.exists()).toBeTruthy();
        expect(textarea.hasClass('resize')).toBeTruthy();
    });
    it('textarea with autocomplete', () => {
        const textAreaComponent = shallow(
            <Input {...defaultInputProps} {...inputTestMockMap.get('textarea-can-be-autocompleted')} />,
        );
        const textarea = textAreaComponent.find('textarea');
        expect(textarea.exists()).toBeTruthy();
        expect(textarea.props().autoComplete).toEqual('on');
    });
    it('input with autocomplete', () => {
        const defaultInput = shallow(
            <Input {...defaultInputProps} {...inputTestMockMap.get('input-can-be-autocompleted')} />,
        );
        const inputWrapper = defaultInput.find('.inputWrapper');
        const input = inputWrapper.find('DebounceInput');
        expect(input.exists()).toBeTruthy();
        expect(input.props().autoComplete).toEqual('on');
    });
    it('render password icon', () => {
        const passwordInput = shallow(
            <Input {...defaultInputProps} {...inputTestMockMap.get('it-render-password-icon')} />,
        );
        const inputWrapper = passwordInput.find('.inputWrapper');
        const passwordIcon = inputWrapper.find('.passwordDisplayButton');
        expect(passwordIcon.exists()).toBeTruthy();
    });
    it('changes eye icon onPointerDown and onPointerUp', () => {
        const passwordInput = shallow(
            <Input {...defaultInputProps} {...inputTestMockMap.get('it-render-password-icon')} />,
        );
        const passwordIcon = passwordInput.find('.inputWrapper').find('.passwordDisplayButton');
        expect(passwordIcon.exists()).toBeTruthy();
        (passwordIcon.props() as { onPointerDown: () => void }).onPointerDown();
        // expect((passwordIcon.props() as { onPointerDown: () => void }).onPointerDown).toHaveBeenCalledTimes(1);
    });
});
