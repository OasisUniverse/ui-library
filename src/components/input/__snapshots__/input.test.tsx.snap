// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> test list render all list 1`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 2`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 3`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 4`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 5`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 6`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 7`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 8`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 9`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 10`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 11`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 12`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 13`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 14`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 15`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 16`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 17`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 18`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 19`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 20`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 21`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 22`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 23`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 24`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 25`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 26`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 27`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 28`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 29`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 30`] = `ShallowWrapper {}`;

exports[`<Input /> test list render all list 31`] = `ShallowWrapper {}`;
